src/storage/src/hummock/sstable_store.rs

```diff
@@ -821,21 +821,66 @@ impl StreamingSstableWriterFactory {
 
 #[async_trait::async_trait]
 impl SstableWriterFactory for StreamingSstableWriterFactory {
-    type Writer = StreamingUploadWriter;
+    type Writer = AnySstableWriter;
 
     async fn create_sst_writer(
         &mut self,
         object_id: HummockSstableObjectId,
         options: SstableWriterOptions,
     ) -> HummockResult<Self::Writer> {
-        let path = self.sstable_store.get_sst_data_path(object_id);
-        let uploader = self.sstable_store.store.streaming_upload(&path).await?;
-        Ok(StreamingUploadWriter::new(
-            object_id,
-            self.sstable_store.clone(),
-            uploader,
-            options,
-        ))
+        if self.sstable_store.store.support_streaming_upload() {
+            let path = self.sstable_store.get_sst_data_path(object_id);
+            let uploader = self.sstable_store.store.streaming_upload(&path).await?;
+            Ok(AnySstableWriter::StreamingUploadWriter(
+                StreamingUploadWriter::new(
+                    object_id,
+                    self.sstable_store.clone(),
+                    uploader,
+                    options,
+                ),
+            ))
+        } else {
+            Ok(AnySstableWriter::BatchSstableWriter(
+                BatchUploadWriter::new(object_id, self.sstable_store.clone(), options),
+            ))
+        }
+    }
+}
+
+pub enum AnySstableWriter {
```

Comment:
I suggest we check `support_streaming_upload` [here](https://github.com/risingwavelabs/risingwave/blob/202e50236d7e78d4c059129c5c76a1948fa7d053/src/storage/src/hummock/compactor/mod.rs#L870) to avoid introducing this enum. It is a bit strange to has a AnySstableWrtier for StreamingSstableWriterFactory because it means that StreamingSstableWrtierFactory can create a sst wrtier that doesn't do streaming upload.

Reply:
Great, thanks!